### PR TITLE
[GLUTEN-4170][VL] Decouple partitions from plan to avoid driver stalled

### PR DESCRIPTION
**Pull request by Yohahaha**
## What changes were proposed in this pull request?

There are two parts will lead driver stalled when scan contains lots of partitions,
1. plan serialization happens in every GlutenPartition construction.
2. GlutenWholeStageColumnarRDD#getPartitions

| |  | 22374 partitions | 44611 partitions |
|--------|--------|--------|--------|
| before | #1 | 880ms | 1352ms |
| | #2 | 3662ms | 17186ms |
| after | #1 | 21ms | 106ms |
|  | #2 | 6ms | 25ms | 

This patch decouple scan splitInfo(LocalFileNodes) from ReadRel to avoid serialize substrait plan for each partition in Driver, when the plan is complex or the number of partitions is particularly large, the cost of this serialization cannot be ignored.

Stream splitInfo(inputIterator) still kept in ReadRel for now.

(Fixes: \#4170)

### CODE REVIEW

**Comment by Yohahaha** [comment]
@FelixYBW @rui-mo @PHILO-HE @ulysses-you 
let's discuss the solution based on current patch, the key is decouple scan partition from plan when serialization. I know this patch may seems tricky, and open for more better design.

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto [hunk 1]
```diff
@@ -159,16 +159,18 @@ message ReadRel {
...
-     repeated partitionColumn partition_columns = 16;
+      }
+      repeated partitionColumn partition_columns = 16;
+
+      /// File schema
+      NamedStruct schema = 17;
 
-     /// File schema
-     NamedStruct schema = 17;
+      bool stream_input = 18;
```
Comment (ulysses-you) [inline gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:173]: is it necessary ? we check if it is a iterator input by comparing the prefix string is equal to iterator

File: gluten-core/src/main/java/io/glutenproject/substrait/rel/LocalFilesNode.java [hunk 2]
```diff
@@ -50,7 +50,7 @@ public enum ReadFileFormat {
   }
 
   private ReadFileFormat fileFormat = ReadFileFormat.UnknownFormat;
-  private Boolean iterAsInput = false;
+  private Boolean streamInput = false;
```
Comment (ulysses-you) [inline gluten-core/src/main/java/io/glutenproject/substrait/rel/LocalFilesNode.java:53]: TBH, the original variable name seems easy to follow. streamInput can be used when Gluten support Flink.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h [hunk 3]
```diff
@@ -33,6 +34,8 @@ struct SplitInfo {
   /// Whether the split comes from arrow array stream node.
   bool isStream = false;
 
+  int32_t idx{-1};
```
Comment (ulysses-you) [inline cpp/velox/substrait/SubstraitToVeloxPlan.h:37]: `iteratorIndex`

**Review by ulysses-you — COMMENTED** [review]
thank you @Yohahaha for the idea, looks reasonable to me. One concern is about the name, can we avoid using `streamXxx` ? I think `iteratorXxx` is easy to follow. We are using iterator in both java and native side.

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto (hunk 1 above)
Comment (Yohahaha) [reply gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto]: it's not necessary, I add it before for debugging purposes, already removed.

File: gluten-core/src/main/java/io/glutenproject/substrait/rel/LocalFilesNode.java (hunk 2 above)
Comment (Yohahaha) [reply gluten-core/src/main/java/io/glutenproject/substrait/rel/LocalFilesNode.java]: fixed.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h (hunk 3 above)
Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.h]: removed.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala [hunk 4]
```diff
@@ -178,6 +210,7 @@ class IteratorApiImpl extends IteratorApi with Logging {
     val nativeResultIterator =
       transKernel.createKernelWithBatchIterator(
         rootNode.toProtobuf.toByteArray,
+        new Array[Array[Byte]](0),
```
Comment (rui-mo) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: Could you explain a bit what happens here, and is split info needed for final iterator?

File: gluten-data/src/main/java/io/glutenproject/vectorized/NativePlanEvaluator.java [hunk 5]
```diff
@@ -61,7 +61,8 @@ public NativePlanValidationInfo doNativeValidateWithFailureReason(byte[] subPlan
   // Used by WholeStageTransform to create the native computing pipeline and
   // return a columnar result iterator.
   public GeneralOutIterator createKernelWithBatchIterator(
-      byte[] wsPlan, List<GeneralInIterator> iterList) throws RuntimeException, IOException {
+      byte[] wsPlan, byte[][] splitInfo, List<GeneralInIterator> iterList)
+      throws RuntimeException, IOException {
```
Comment (rui-mo) [inline gluten-data/src/main/java/io/glutenproject/vectorized/NativePlanEvaluator.java:65]: Could we ensure `splitInfo` is not empty?

**Review by rui-mo — COMMENTED** [review]
Thanks for the improvement.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 4 above)
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: SplitInfo only exists when stage contain scan, so the final iterator does not need it I think.

`new Array[Array[Byte]](0)` is an empty byte array which represents empty split infos, I know it's looks wired, maybe use a new proto class for array of split infos would be better, but needs change lots codes of substrait conversion. How do you think about that?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala [hunk 6]
```diff
@@ -128,13 +145,28 @@ class IteratorApiImpl extends IteratorApi with Logging {
...
     val columnarNativeIterators =
       new JArrayList[GeneralInIterator](inputIterators.map {
         iter => new ColumnarBatchInIterator(iter.asJava)
       }.asJava)
     val transKernel = NativePlanEvaluator.create()
+
+    val splitInfoByteArray = inputPartition
+      .asInstanceOf[GlutenPartition]
+      .splitInfos
+      .map(_.asInstanceOf[LocalFilesNode].toProtobuf.toByteArray)
```
Comment (ulysses-you) [inline backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:162]: shall we call LocalFilesNode to byte array at driver side ?

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 7]
```diff
@@ -809,21 +844,20 @@ core::PlanNodePtr SubstraitToVeloxPlanConverter::toVeloxPlan(const ::substrait::
...
+    return constructValueStreamNode(readRel, streamIdx);
   }
 
   // Otherwise, will create TableScan node for ReadRel.
+  auto splitInfo = std::make_shared<SplitInfo>();
+  if (!validationMode_) {
+    VELOX_CHECK_LT(splitInfoIdx_, splitInfos_.size(), "Plan must have readRel and related split info.");
+    splitInfo = splitInfos_[splitInfoIdx_++];
+  }
+  fileFormat_ = splitInfo->format;
```
Comment (ulysses-you) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:859]: it seems `fileFormat_` is unused

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 8]
```diff
@@ -809,21 +844,20 @@ core::PlanNodePtr SubstraitToVeloxPlanConverter::toVeloxPlan(const ::substrait::
         !readRel.common().has_emit(), "Emit not supported for ValuesNode and TableScanNode related Substrait plans.");
   }
 
-  // Check if the ReadRel specifies an input of stream. If yes, the pre-built
-  // input node will be used as the data source.
-  auto splitInfo = std::make_shared<SplitInfo>();
+  // Check if the ReadRel specifies an input of stream. If yes, build ValueStreamNode as the data source.
   auto streamIdx = getStreamIndex(readRel);
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:848]: `getStreamIndex` relies on local files to extract the iterator index. Could you help confirm?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 4 above)
Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: I'm wondering if split info is also needed for final iterator if ReadRel does not contain LocalFiles. See https://github.com/oap-project/gluten/pull/4177#discussion_r1439213063.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc (hunk 8 above)
Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:848]: yes.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 4 above)
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: scan split info is not needed for final iterator, but stream split info is needed.

File: backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala [hunk 9]
```diff
@@ -95,6 +90,24 @@ class CHIteratorApi extends IteratorApi with Logging with LogLevelUtil {
...
+   * Generate native row partition.
+   *
+   * @return
+   */
+  override def genPartitions(wsCtx: WholeStageTransformContext, splitInfos: Seq[Seq[SplitInfo]]): Seq[BaseGlutenPartition] = {
+    splitInfos.zipWithIndex.map {
+      case (splitInfos, index) =>
+        wsCtx.substraitContext.initSplitInfosIndex(0)
+        wsCtx.substraitContext.setSplitInfos(splitInfos)
+        val substraitPlan = wsCtx.root.toProtobuf
```
Comment (PHILO-HE) [inline backends-clickhouse/src/main/scala/io/glutenproject/backendsapi/clickhouse/CHIteratorApi.scala:103]: I'm assuming the proposed optimization can also be applied for CH backend. If so, it will need some follow-up work from CH engineer. @baibaichen

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 6 above)
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:162]: good point, I will check later.

**Comment by Yohahaha** [comment]
I assume this solution looks well for all your guys.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 4 above)
Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: Does this PR cover this change or some extra change is needed?

Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:213]: > Does this PR cover this change or some extra change is needed?

this patch does change stream split info's behavior, it will be extracted from ReadRel as before.

**Comment by zhouyuan** [comment]
@lgbo-ustc hi, this patch tries to refactor on gen file partitions, please take a look if it will impact CK backend

thanks, -yuan

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc (hunk 7 above)
Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc]: yes.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala [hunk 10]
```diff
@@ -80,6 +76,27 @@ class IteratorApiImpl extends IteratorApi with Logging {
     }
   }
 
+  /**
+   * Generate native row partition.
+   *
+   * @return
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:82]: removed.

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto [hunk 11]
```diff
@@ -159,16 +159,16 @@ message ReadRel {
...
         string value = 2;
-     }
-     repeated partitionColumn partition_columns = 16;
+      }
+      repeated partitionColumn partition_columns = 16;
 
-     /// File schema
-     NamedStruct schema = 17;
+      /// File schema
+      NamedStruct schema = 17;
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:171]: it's needed, there should be 6 leading space.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 12]
```diff
@@ -1157,10 +1149,7 @@ int32_t SubstraitToVeloxPlanConverter::getStreamIndex(const ::substrait::ReadRel
       VELOX_FAIL(err.what());
     }
   }
-  if (validationMode_) {
-    return -1;
-  }
-  VELOX_FAIL("Local file is expected.");
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:1281]: ReadRel may not have local file node when it's a real scan node.

**Comment by Yohahaha** [comment]
I see QueryBenchmark is similar with GenericBenchmark and QueryBenchmark does not covered by CI, could we remove it? @marin-ma @jinchengchenghh

**Comment by marin-ma** [comment]
> I see QueryBenchmark is similar with GenericBenchmark and QueryBenchmark does not covered by CI, could we remove it? @marin-ma @jinchengchenghh

cc: @rui-mo

**Comment by rui-mo** [comment]
> I see QueryBenchmark is similar with GenericBenchmark and QueryBenchmark does not covered by CI, could we remove it?

@Yohahaha GenericBenchmark uses arrow to read files, while QueryBenchmark uses Velox. So QueryBenchmark is useful when we want to test Velox TableScan. I think the better option is to enable QueryBenchmark on CI. @marin-ma Please help to confirm, thanks.

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala (hunk 4 above)
Comment (rui-mo) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:200]: > Final iterator does not need split infos, and an empty byte array is used to represent empty split infos.

Maybe better to add some comments as above, thanks.

File: cpp/velox/benchmarks/GenericBenchmark.cc [hunk 13]
```diff
@@ -126,7 +127,9 @@ auto BM_Generic = [](::benchmark::State& state,
   auto memoryManager = getDefaultMemoryManager();
   auto runtime = Runtime::create(kVeloxRuntimeKind, conf);
   const auto& filePath = substraitJsonFile;
-  auto plan = getPlanFromFile(filePath);
+  auto plan = getPlanFromFile("Plan", filePath);
+  const auto& splitPath = splitFile;
```
Comment (rui-mo) [inline cpp/velox/benchmarks/GenericBenchmark.cc:131]: Why do we need this variable, can we use splitFile instead?

File: cpp/velox/compute/VeloxPlanConverter.h [hunk 14]
```diff
@@ -35,45 +35,19 @@ class VeloxPlanConverter {
       const std::unordered_map<std::string, std::string>& confMap,
       bool validationMode = false);
 
-  std::shared_ptr<const facebook::velox::core::PlanNode> toVeloxPlan(::substrait::Plan& substraitPlan);
+  std::shared_ptr<const facebook::velox::core::PlanNode> toVeloxPlan(
+      ::substrait::Plan& substraitPlan,
```
Comment (rui-mo) [inline cpp/velox/compute/VeloxPlanConverter.h:39]: Not your change. Maybe better to add const here.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h [hunk 15]
```diff
@@ -137,6 +140,15 @@ class SubstraitToVeloxPlanConverter {
     planNodeId_ = planNodeId;
   }
 
+  void setSplitInfos(std::vector<std::shared_ptr<SplitInfo>> splitInfos) {
+    splitInfoIdx_ = 0;
```
Comment (rui-mo) [inline cpp/velox/substrait/SubstraitToVeloxPlan.h:144]: > int32_t splitInfoIdx_{0};

Why do we need to set split info index as 0? Its initial value should already be 0.

**Comment by marin-ma** [comment]
> > I see QueryBenchmark is similar with GenericBenchmark and QueryBenchmark does not covered by CI, could we remove it?
> 
> @Yohahaha GenericBenchmark uses arrow to read files, while QueryBenchmark uses Velox. So QueryBenchmark is useful when we want to test Velox TableScan. I think the better option is to enable QueryBenchmark on CI. @marin-ma Please help to confirm, thanks.

@rui-mo If input is from middle stage, GenericBenchmark will use arrow reader to load the input iterator. If input is from first stage, the whole pipeline is offloaded including table scan. Here's the doc https://github.com/oap-project/gluten/blob/main/docs/developers/MicroBenchmarks.md#generate-substrait-plan-and-input-for-any-query

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto (hunk 11 above)
[the remark this answers is not in the document]
Comment (rui-mo) [reply gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:171]: Is it about the format? How do you meet these errors?

**Comment by rui-mo** [comment]
@marin-ma Thanks for confirming. @Yohahaha We can remove QueryBenchmark because its functionality is covered by `GenericBenchmark`.

**Comment by FelixYBW** [comment]
> @marin-ma Thanks for confirming. @Yohahaha We can remove QueryBenchmark because its functionality is covered by `GenericBenchmark`.

Agree

File: cpp/velox/benchmarks/GenericBenchmark.cc (hunk 13 above)
Comment (Yohahaha) [reply cpp/velox/benchmarks/GenericBenchmark.cc:131]: just follow `filePath`, will remove.

File: cpp/velox/compute/VeloxPlanConverter.h (hunk 14 above)
Comment (Yohahaha) [reply cpp/velox/compute/VeloxPlanConverter.h:40]: ok.

File: cpp/velox/substrait/SubstraitToVeloxPlan.h (hunk 15 above)
Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.h:145]: make sure each call to `setSplitInfos` will reset `splitInfoIdx_` correctly.

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 16]
```diff
@@ -35,183 +34,87 @@ VeloxPlanConverter::VeloxPlanConverter(
...
   }
-}
 
-void VeloxPlanConverter::setInputPlanNode(const ::substrait::FilterRel& sfilter) {
-  if (sfilter.has_input()) {
-    setInputPlanNode(sfilter.input());
-  } else {
-    throw std::runtime_error("Child expected");
-  }
+  planConverter->setSplitInfos(std::move(splitInfos));
```
Comment (rui-mo) [inline cpp/velox/compute/VeloxPlanConverter.cc:107]: Will we call `setSplitInfos` multiple times for an instance of planConverter?
https://github.com/oap-project/gluten/pull/4177#discussion_r1441506167

**Comment by yma11** [comment]
@Yohahaha Can you rebase this PR? then I think the CI/CD should can pass.

**Comment by Yohahaha** [comment]
> @Yohahaha Can you rebase this PR? then I think the CI/CD should can pass.

I will request review&&merge when I fix all ci issue, thanks.

File: gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto (hunk 11 above)
[the remark this answers is not in the document]
Comment (Yohahaha) [reply gluten-core/src/main/resources/substrait/proto/substrait/algebra.proto:171]: Just found this format issue and fix it.

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 16 above)
Comment (Yohahaha) [reply cpp/velox/compute/VeloxPlanConverter.cc:106]: should not, but not sure.

**Comment by Yohahaha** [comment]
I believe CI could pass, any more comments? cc who still watching this patch.

**Comment by FelixYBW** [comment]
looks the micro benchmark failed, @marin-ma any insight?

**Comment by Yohahaha** [comment]
> looks the micro benchmark failed, @marin-ma any insight?

I will fix it later.

**Comment by Yohahaha** [comment]
please help review again, CI has passed. @FelixYBW @rui-mo @PHILO-HE @ulysses-you

File: gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala [hunk 17]
```diff
@@ -429,6 +419,9 @@ class ColumnarInputRDDsWrapper(columnarInputRDDs: Seq[RDD[ColumnarBatch]]) exten
   }
 
   def getPartitions(index: Int): Seq[Partition] = {
+    if (columnarInputRDDs.isEmpty) {
+      return Seq.empty
+    }
```
Comment (ulysses-you) [inline gluten-core/src/main/scala/io/glutenproject/execution/WholeStageTransformer.scala:424]: is this change necessary ?

File: cpp/core/jni/JniWrapper.cc [hunk 18]
```diff
@@ -384,7 +385,16 @@ Java_io_glutenproject_vectorized_PlanEvaluatorJniWrapper_nativeCreateKernelWithI
   auto planData = reinterpret_cast<const uint8_t*>(env->GetByteArrayElements(planArr, nullptr));
   auto planSize = env->GetArrayLength(planArr);
 
+  if (jsize splitInfoArraySize = env->GetArrayLength(splitInfosArr); splitInfoArraySize != 0) {
```
Comment (ulysses-you) [inline cpp/core/jni/JniWrapper.cc:388]: move these code to line 384

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 19]
```diff
@@ -34,183 +33,87 @@ VeloxPlanConverter::VeloxPlanConverter(
...
-  if (sagg.has_input()) {
-    setInputPlanNode(sagg.input());
-  } else {
-    throw std::runtime_error("Child expected");
-  }
-}
+void parseLocalFileNodes(
+    SubstraitToVeloxPlanConverter* planConverter,
+    std::vector<::substrait::ReadRel_LocalFiles>& localFiles) {
+  std::vector<std::shared_ptr<SplitInfo>> splitInfos;
```
Comment (ulysses-you) [inline cpp/velox/compute/VeloxPlanConverter.cc:98]: `splitInfos.reserve(localFiles.size())`

File: cpp/velox/compute/VeloxPlanConverter.cc [hunk 20]
```diff
@@ -34,183 +33,87 @@ VeloxPlanConverter::VeloxPlanConverter(
...
+  // avoid include RowVectorStream.h in SubstraitToVeloxPlan.cpp, it may cause redefintion of array abi.h.
+  auto factory = [](std::string nodeId,
+                    memory::MemoryPool* pool,
+                    std::shared_ptr<ResultIterator> inputIter,
+                    RowTypePtr outputType) {
+    auto vectorStream = std::make_shared<RowVectorStream>(pool, std::move(inputIter), outputType);
+    return std::make_shared<ValueStreamNode>(nodeId, outputType, std::move(vectorStream));
+  };
+  substraitVeloxPlanConverter_.setValueStreamNodeFactory(std::move(factory));
+  substraitVeloxPlanConverter_.setInputIterators(inputIters);
```
Comment (zhztheplayer) [inline cpp/velox/compute/VeloxPlanConverter.cc:48]: If using this way, should we remove `setInputIterators` then capture `inputIters` in factory body?

File: backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala [hunk 21]
```diff
@@ -80,6 +76,24 @@ class IteratorApiImpl extends IteratorApi with Logging {
...
+      splitInfos: Seq[Seq[SplitInfo]]): Seq[BaseGlutenPartition] = {
+    // Only serialize plan once, save lots time when plan is complex.
+    val planByteArray = wsCtx.root.toProtobuf.toByteArray
+
+    splitInfos.zipWithIndex.map {
+      case (splitInfos, index) =>
+        GlutenPartition(
+          index,
+          planByteArray,
+          splitInfos.map(_.asInstanceOf[LocalFilesNode].toProtobuf.toByteArray).toArray,
```
[the remark this answers is not in the document]
Comment (Yohahaha) [reply backends-velox/src/main/scala/io/glutenproject/backendsapi/velox/IteratorApiImpl.scala:91]: yes.

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 22]
```diff
@@ -1297,7 +1286,7 @@ int32_t SubstraitToVeloxPlanConverter::getStreamIndex(const ::substrait::ReadRel
     std::string filePath = fileList[0].uri_file();
     std::string prefix = "iterator:";
     std::size_t pos = filePath.find(prefix);
-    if (pos == std::string::npos) {
+    if (pos == std::string::npos || validationMode_) {
```
Comment (zhztheplayer) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:1289]: Can we let the function return `-1` fast in `validationMode_` ?

E.g., 
```cpp
if (validationMode_) {
  return -1;
}
```

At the entrance of the function.

cc @rui-mo

**Review by zhztheplayer — COMMENTED** [review]
Thanks for working on this!

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 20 above)
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxPlanConverter.cc:48]: Sorry, I still don't quite understand the point here. Capturing `inputIters` won't make the factory stateful since once it is created, nobody could change its internal creation logic of the value stream node. And e could even eliminate stateful `setInputIterators` by the change as well. Any special reason you wanted to keep it?

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc (hunk 22 above)
Comment (zhztheplayer) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:1289]: It should effectively be the same with the patch's code
```cpp
if (pos == std::string::npos || validationMode_) {
  return -1;
}
```

The code looks to return -1 when `validationMode_ == true` anyway?

Comment (Yohahaha) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc]: you are right, thanks!

File: cpp/velox/compute/VeloxPlanConverter.cc (hunk 20 above)
Comment (Yohahaha) [reply cpp/velox/compute/VeloxPlanConverter.cc]: got it, thanks!

**Review by ulysses-you — DISMISSED** [review]
lgtm, thank you @Yohahaha

**Review by marin-ma — CHANGES_REQUESTED** [review]
Thank you for this work! Could you please also update the micro benchmark documentation? Noticed that we need to specify split files for first stages.

**Review by marin-ma — APPROVED** [review]

**Comment by marin-ma** [comment]
LGTM. Thanks!

**Review by ulysses-you — APPROVED** [review]

**Comment by ulysses-you** [comment]
We can create followups if there are some new finding issues

File: cpp/velox/substrait/SubstraitToVeloxPlan.cc [hunk 23]
```diff
@@ -960,6 +960,36 @@ core::PlanNodePtr SubstraitToVeloxPlanConverter::toVeloxPlan(const ::substrait::
   }
 }
 
+core::PlanNodePtr SubstraitToVeloxPlanConverter::constructValueStreamNode(
+    const ::substrait::ReadRel& readRel,
+    int32_t streamIdx) {
+  // Get the input schema of this iterator.
+  uint64_t colNum = 0;
+  std::vector<TypePtr> veloxTypeList;
+  if (readRel.has_base_schema()) {
```
Comment (jinchengchenghh) [inline cpp/velox/substrait/SubstraitToVeloxPlan.cc:969]: I find here you change the input iterator names, and comment ValueStreamNode in Velox does not support name change, but here updates the output type based on output type, do you meet any problem? @Yohahaha 
https://github.com/apache/incubator-gluten/blob/c5b7e59335201f960cda49dff7edc315b36ed05e/cpp/velox/operators/plannodes/RowVectorStream.h#L100

Comment (jinchengchenghh) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:969]: And this update only take effect on the top level columns, what if there is nested column?

Comment (jinchengchenghh) [reply cpp/velox/substrait/SubstraitToVeloxPlan.cc:969]: I know this is used for constructing ValueStreamNode outputType, may also need to update nested column
